### ISSUE TITLE
x/net/http2: support consuming PUSH_PROMISE streams in the client

### ISSUE THREAD

**Issue by glasser** [golang/go#18594]
Go 1.8 will contain support for servers to produce PUSH_PROMISE streams, but there is no support for clients to receive them.

Specifically, it would be nice if proxies built with net/http/httputil's ReverseProxy to be able to proxy server pushes.

**Comment by bradfitz** [comment]
Let's ignore ReverseProxy for the purpose of this bug. Although, ReverseProxy being in std would require that new API be added to net/http.Client.

I think more immediately we'd want to add experimental, opt-in API only to golang.org/x/net/http2.Transport. (Unless opted-in, the Transport would continue to advertise SETTINGS_ENABLE_PUSH == 0)

Got any API proposals?

**Comment by glasser** [comment]
I admit I'm not an http2 expert. I maintain a proxy based on ReverseProxy and "proxy http2 server pushes" is a feature request I got from a user.

My understanding is that receiving a push is effectively similar to receiving a request and a response from a server.  So perhaps there's a new field on http2.Transport which is a ReceivePusher, where

```
type ReceivePusher interface {
  ReceivePush(*http.Request, *http.Response) error
}
```

If the field is non-nil, then the Transport advertises SETTINGS_ENABLE_PUSH == 1, and upon receiving a PUSH_PROMISE frame, the Transport creates a Request and a Response that match the PUSH_PROMISE and passes it to the ReceivePusher.  Just like with RoundTrip, it is the responsibility of ReceivePush to read from and Close the response's Body.

I'm not sure what is the best way to indicate "I don't want this push" (ie send a RST_STREAM I think?) — maybe any non-nil error return value?  Maybe a second return value other than the error?  Some other interface passed in that it can invoke? Some method on the Transport?

My understanding is that these pushes can be received at any time during the connection: concurrently with RoundTrip or after it.  Not really sure how that should affect the API.

**Comment by tombergan** [comment]
It seems right to translate each PUSH_PROMISE into an http.Request that has no request body. A common thing in browsers is to inspect the PUSH_PROMISE, check if it's already cached, and if so, cancel the push with RST_STREAM(CANCEL). Whatever the API looks like, it should allow canceling the pushed stream _before_ any bytes of the response have been received.

A second point is that every PUSH_PROMISE must be associated with a client request. This is required by RFC 7540. It may be useful to communicate that associated request to the client's push handler.

**Comment by nilslice** [comment]
I think the interface proposed by @glasser is good, though would likely need some additional control methods, for example to cancel/block the PUSH_PROMISE frame. as per h2 spec:
>If the client determines, for any reason, that it does not wish to receive the pushed response from the server or if the server takes too long to begin sending the promised response, the client can send a RST_STREAM frame, using either the CANCEL or REFUSED_STREAM code and referencing the pushed stream's identifier.

To keep it limited to the http2.Transport suggested by @bradfitz, it would seem like some configuration to the RoundTripOpt would be required to state whether the Transport.RoundTrip Response should check for some cache of stream IDs, but I'm not sure where it makes sense to do the bookkeeping.

The ReceivePush method could return an error signaling the decision CANCEL/REFUSE_STREAM, but how it comes to that conclusion should be based on some configurable rules. If it accepts the push, does the Transport keep track of the stream IDs to read later?

**Comment by rayvbr** [comment]
Looks good to me. If there's anything I can do to help this along, let me know.

**Comment by tombergan** [comment]
Once @bradfitz is happy with the interface, we'd be happy for a volunteer to do the implementation. I won't have a chance to do this for a few weeks, at least, and I'm not sure if Brad will either.

**Comment by rayvbr** [comment]
Although I'm not sure whether my expertise of both the innards of x/net/http as well as the HTTP/2 spec is sufficient to bring this task to completion, with some guidance I'm definitely willing to give it a shot.

**Comment by snadrus** [comment]
I'd like this to be a way for grpc have a server-initiated rpc call for connected clients by using this with an http call for the response. Then grpc covers this significant use case that previously required websocket.

**Comment by bradfitz** [comment]
@snadrus, what does grpc have to do with PUSH_PROMISE? I see nothing in http://www.grpc.io/docs/guides/wire.html

**Comment by snadrus** [comment]
Nothing currently, however server-initiated rpc calls are missing from grpc, so enabling this through push frames may be an option. I'm investigating this with the grpc team.

**Comment by tombergan** [comment]
@snadrus does your desired use in grpc require specific API features? If so, do you believe the API suggested above will work, or do you have other suggestions for an API?

At this point, we're more interested in API suggestions and critiques than "me toos" (we realize this is a desired feature!).

**Comment by romainmenke** [comment]
@glasser For a proxy it might be interesting to look at the Cloudflare approach : [CF Push Blog](https://blog.cloudflare.com/announcing-support-for-http-2-server-push-2/)

I wrote an `http.Handler` wrapper to enable this, it has been tested extensively but we aren't using this in production yet.
[github.com/romainmenke/pusher/tree/master/link](https://github.com/romainmenke/pusher/tree/master/link)

I would also be very interested in helping out in any way I can with a client implementation as a push capable client would be a great help when writing tests for pushes from a server.

**Comment by ajackson-cpi** [comment]
@tombergan Yes, this works for the use I brought up.

**Comment by captncraig** [comment]
I would like client support for push in order to write automated verification tests for our http2 infrastructure. I am not very familiar with the mechanics of push, but I am really mostly interested in knowing that the pushes are initiated properly. I don't even need to read them, although that would be bonus.

**Comment by twdkeule** [comment]
Is there a use-case where we need the `OriginalRequest`? Because in normal operation the client does not really care where the resource comes from. 

Would `ReadResponse` actually make the flow window decrease (reading from the stream) or would there be some kind of buffering so the connection does not become flow limited if no one is reading this pushed resource?

**Comment by romainmenke** [comment]
Maybe it would be best to also use `context.Context` instead `http.Request` as a way to identify what triggered the push, just to avoid implementations like this : #20948

I already opened an issue to add this information on the server side : #20566

**Comment by brunomcustodio** [comment]
Is there any chance that this will make it to Go1.9?

**Comment by bradfitz** [comment]
@brunomcustodio, zero chance. There's not even code, and also https://golang.org/wiki/Go-Release-Cycle

**Comment by twdkeule** [comment]
Any update on this?

**Comment by bradfitz** [comment]
Nope.

**Comment by meirf** [comment]
@nilslice @romainmenke looks like y'all have h2 push implementation experience. Any interest in giving feedback on https://golang.org/cl/85577? Even partial feedback/minor comments would be helpful.

**Comment by himulawang** [comment]
Any update? We really need this feature.

**Comment by agnivade** [comment]
@himulawang - Please take a look at https://github.com/golang/go/wiki/NoPlusOne.

**Comment by the729** [comment]
@meirf I see the review marathon on your https://golang.org/cl/85577. Are you still interested in working on it? I'm happy to help implementing/merging/working on this issue. 
Do you have a public repo that contains your patch commits?

**Comment by the729** [comment]
@meirf I also feel this is a change too large to be merged once. Probably split into Framer + Transport. It seems the Transport API is still experimental, but changes to Framer is far less controversial.

**Comment by meirf** [comment]
@the729 

> Are you still interested in working on it?

I don't have the bandwidth to work on this for the foreseeable future. I welcome someone to take it over. (My only request is that the original CL url https://golang.org/cl/85577 be included in the commit message of new CLs.)

> I also feel this is a change too large to be merged once. Probably split into Framer + Transport. It seems the Transport API is still experimental, but changes to Framer is far less controversial.

I agree that splitting it up starting with the less controversial changes is a great strategy and probably what I should have done. 👍

**Comment by the729** [comment]
I will take it over from @meirf to work on this issue, and I have created a new CL https://golang.org/cl/181497.

Since @bradfitz is on leave recently, is there anyone else interested in reviewing this CL?

**Comment by the729** [comment]
ping

**Comment by bradfitz** [comment]
@the729, sorry for the delay. I'm back at work full time now but still working through a backlog of stuff. I took a quick look at the CL and left one comment. I haven't dug into the code much, but overall seems reasonable.

**Comment by Richard87** [comment]
Hi! I'm really looking forward for this being implemented, what can I do to help?

I'm a user of Traefik that uses h2... (https://github.com/containous/traefik/issues/906)

**Comment by the729** [comment]
The CL is still waiting for @bradfitz and others' review: https://golang.org/cl/181497

**Comment by phanirithvij** [comment]
@bradfitz gentle ping

**Comment by the729** [comment]
I'm glad there are people still interested in this feature. I'll probably resolve the merge conflicts some day, making the CL fresh and shiny again and ready to be reviewed....

**Comment by BRUHItsABunny** [comment]
ping

In all seriousness though... Is there anything we, the community, can do to move this along?

**Comment by ianlancetaylor** [comment]
CC @neild

**Comment by neild** [comment]
My understanding is that HTTP server push is largely unused at this point, and Chromium is planning on dropping support for the feature entirely:
https://groups.google.com/a/chromium.org/g/blink-dev/c/K3rYLvmQUBY/m/vOWBKZGoAQAJ?pli=1

Is this a feature we should be adding at this point? Or is it already obsolete?

**Comment by mitar** [comment]
I think similar to what others raised in that Chromium mailing list thread: it is hard to argue that HTTP server push is largely unused if it has not been fully available to begin with. For example, this issue is open since 2017. So yea, it is a bit of chicken and egg problem. Unless support is there, it is hard to make use of it. And this happens in various places along the stack.

So I would still ask for this to be implemented. In my case: it is easier to write unit tests against a server which does implement HTTP2 push.

**Comment by bradfitz** [comment]
I ran into this again and forgot it had stalled. Chatting with @neild now about this. We're moving it into the proposal process. LGTM overall, though.

**Comment by rsc** [comment]
Can someone summarize what the proposed new API is? Is it the API in https://go-review.git.corp.google.com/c/net/+/181497?
(If so can you summarize that?)

**Comment by rsc** [comment]
This proposal has been added to the [active column](https://go.dev/s/proposal-status#active) of the proposals project
and will now be reviewed at the weekly proposal review meetings.
— rsc for the proposal review group

**Comment by rsc** [comment]
I think we are still waiting for someone to summarize the new API.

**Comment by the729** [comment]
```golang
package http2

// Transport is an HTTP/2 Transport.
//
// A Transport internally caches connections to servers. It is safe
// for concurrent use by multiple goroutines.
type Transport struct {
...
...
	// PushHandler is called upon receiving PUSH_PROMISEs from the server.
	// If nil, server push is disabled.
	//
	// Unless TLSClientConfig.InsecureSkipVerify is set, the Transport verifies
	// whether the server is authoritative for the received PUSH_PROMISEs. On a
	// TLS connection, this is done by verifing the certificates. On a non-TLS
	// connection, the pushed request must have the same host name as the
	// original one.
	//
	// There is no support for limiting the number of responses
	// that can be concurrently pushed by the server, for example, by setting
	// SETTINGS_MAX_CONCURRENT_STREAMS.
	PushHandler PushHandler
...
...
}

// PushHandler consumes a pushed response.
type PushHandler interface {
	// HandlePush will be called once for every PUSH_PROMISE received
	// from the server. If HandlePush returns before the pushed stream
	// has completed, the pushed stream will be canceled.
	HandlePush(r *PushedRequest)
}

// PushedRequest describes a request that was pushed from the server.
type PushedRequest struct {
	// Promise is the HTTP/2 PUSH_PROMISE message. The promised
	// request does not have a body. Handlers should not modify Promise.
	//
	// Promise.RemoteAddr is the address of the server that started this push request.
	Promise *http.Request

	// OriginalRequestURL is the URL of the original client request that triggered the push.
	OriginalRequestURL *url.URL

	// OriginalRequestHeader contains the headers of the original client request that triggered the push.
	OriginalRequestHeader http.Header

	pushedStream *clientStream
}

// ReadResponse reads the pushed response. If ctx is done before the
// response headers are fully received, ReadResponse will fail and PushedRequest
// will be cancelled.
func (pr *PushedRequest) ReadResponse(ctx context.Context) (*http.Response, error) {
        ...
}

// Cancel tells the server that the pushed response stream should be terminated.
// See: https://tools.ietf.org/html/rfc7540#section-8.2.2
func (pr *PushedRequest) Cancel() {
        ...
}
```

The following code demonstrates how to use it.
```golang
client := http.Client{
    Transport: &http2.Transport{
        PushHandler: &somePushHandler{},
    },
}

resp, _ := client.Get(url)
```

**Comment by rsc** [comment]
Thanks @the729. Does anyone object to this API?

@neild, @bradfitz, does it look good?

**Comment by neild** [comment]
This API seems reasonable to me.

**Comment by rsc** [comment]
Based on the discussion above, this proposal seems like a **[likely accept](https://go.dev/s/proposal-status#likely-accept)**.
— rsc for the proposal review group

**Comment by rsc** [comment]
No change in consensus, so **[accepted](https://go.dev/s/proposal-status#accepted)**. 🎉
This issue now tracks the work of implementing the proposal.
— rsc for the proposal review group

**Comment by Chenpinji** [comment]
Does anybody know the situation of the reverse proxy to handle a http/2 push stream? I have tested some popular reverse proxies, but all of them disabled the PUSH PROMISE STREAM. In this topic, it seems that Go doesn't have a module to support a reverse proxy to **handle PUSH PROMISE STREAM**. What is going on now? How about other programming languages or implementations? It should note that my question is about proxy forwarding h2 pushes, I know how to use Link Header to enable server push, but that's not what I want.

**Comment by the729** [comment]
Updated: https://go-review.googlesource.com/c/net/+/181497
